From 79e3698f782b6cf7ac46488fbca140d8d2d91505 Mon Sep 17 00:00:00 2001
From: Xiaowen Liu <b37945@freescale.com>
Date: Mon, 18 Mar 2013 14:26:29 +0800
Subject: [PATCH 431/486] ENGR00253673 fix system hang when recording with GPU
 2d acceleration.

GPU 2D context is combined with a TLS.
so, the gpu 2d hanle must opened and used in the same thread.
otherwise, it will use wrong GPU 2D context and cause system hang.

Signed-off-by: Xiaowen Liu <b37945@freescale.com>
---
 mx6/libcamera2/StreamAdapter.cpp | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/mx6/libcamera2/StreamAdapter.cpp b/mx6/libcamera2/StreamAdapter.cpp
index 5dced50..0d003cd 100755
--- a/mx6/libcamera2/StreamAdapter.cpp
+++ b/mx6/libcamera2/StreamAdapter.cpp
@@ -22,14 +22,10 @@ StreamAdapter::StreamAdapter(int id)
       mMaxProducerBuffers(0), mNativeWindow(NULL), mStreamState(STREAM_INVALID), mReceiveFrame(true)
 {
     g2dHandle = NULL;
-    g2d_open(&g2dHandle);
 }
 
 StreamAdapter::~StreamAdapter()
 {
-    if (g2dHandle != NULL) {
-        g2d_close(g2dHandle);
-    }
 }
 
 int StreamAdapter::initialize(int width, int height, int format, int usage, int bufferNum)
@@ -181,6 +177,10 @@ bool StreamAdapter::handleStream()
                 mStreamState = STREAM_STARTED;
             }
 
+            if (g2dHandle == NULL) {
+                g2d_open(&g2dHandle);
+            }
+
             break;
 
         case STREAM_STOP:
@@ -192,6 +192,11 @@ bool StreamAdapter::handleStream()
                 mStreamState = STREAM_STOPPED;
             }
 
+            if (g2dHandle != NULL) {
+                g2d_close(g2dHandle);
+                g2dHandle = NULL;
+            }
+
             break;
 
         case STREAM_EXIT:
-- 
1.8.0

